src/engines/pallet_engine/mod.rs

```diff
@@ -0,0 +1,601 @@
+		Ok(())
+	}
+	/// Insert string at line. Errors if line number doesn't exist in `output`
+	/// cursor should be handled by caller
+	fn insert_at(&self, line: usize, str: &str) -> anyhow::Result<()> {
+		let reader = BufReader::new(File::open(&self.output)?);
+		let mut temp_file = tempfile::NamedTempFile::new()?;
+		let temp_path = &temp_file.path().to_path_buf();
+		let mut written = false;
+		for (idx, l) in reader.lines().enumerate() {
+			if idx == line {
+				writeln!(temp_file, "{str}")?;
+				written = true;
+			}
+			writeln!(temp_file, "{}", l?)?;
+		}
+		fs::rename(temp_path, &self.output)?;
+		if !written {
+			bail!("PalletEngine output doesn't have line {line} to insert at");
+		}
+		Ok(())
+	}
+	/// Insert raw string at construct_runtime! Fails if there's no construct_runtime! in `output`
+	/// By default inserts at the end of the macro.
+	/// Note: cursor is only incremented by one, so `str` is expected to be a single line
+	fn insert_str_runtime(&mut self, str: &str) -> anyhow::Result<()> {
+		let runtime_contents = fs::read_to_string(&self.output)?;
+		let ast: syn::File = syn::parse_file(runtime_contents.as_ref())?;
+		let mut runtime_found = false;
+		for item in ast.items.iter() {
+			match item {
+				syn::Item::Macro(syn::ItemMacro { mac, .. }) => {
+					if let Some(mac_id) = mac.path.get_ident() {
+						if mac_id == "construct_runtime" {
+							runtime_found = true;
+							let r: RuntimeDeclaration = mac.parse_body().map_err(|err| {
+								anyhow!("PalletEngine output is in not parseable").context(err)
+							})?;
+							match r {
+								RuntimeDeclaration::Implicit(i) => {
+									let ultimate = i
+										.pallets
+										.last()
+										// This may be handled in the case that there's no pallets in CRT to begin with
+										// And subsequent pallets will be added
+										.expect("Fatal: No pallets defined in construct_runtime!")
+										.clone();
+									let end = ultimate.span.end().line;
+									self.insert_at(end, str)?;
+								},
+								RuntimeDeclaration::Explicit(e)
+								| RuntimeDeclaration::ExplicitExpanded(e) => {
+									let end = e
+										.pallets
+										.last()
+										.expect("Fatal: No pallets defined in construct_runtime!")
+										.span
+										.end()
+										.line;
+									self.insert_at(end, str)?;
+								},
+							}
+						}
+					}
+				},
+				_ => {},
+			}
+		}
+		if !runtime_found {
+			// Should never happen
+			panic!("Construct Runtime not found in PalletEngine output. Cannot add pallet");
+		}
+		// Typically insert_str_runtime would be used to insert a single pallet entry so this is okay
+		self.cursor += 1;
+		Ok(())
+	}
+	/// Add a new pallet to RuntimeDeclaration and return it.
+	/// Used to pass a typed AddPallet to modify the RuntimeDeclaration
+	/// Overwrites existing CRT in output, this means that `self.cursor` has to backtracked to crt_start
+	/// and updated after the pallet entry has been inserted
```

Comment:
Sorry but I simply don't understand what you are saying here :)

In general, you explain what is happening in the code very detailed, well done! However, it would make the reviewer's work, and anyone diving into the code, if you could also add a high level explanation of what the code / function does. As a result, the reviewer can get a high level understanding of what the function is doing / meant to do before going going over the code in detail.

It would be really nice if you could document your code in a way where:
1. First, you explain high level what this function does, or what the code from now on (when following the code) is about to do. in a way where a (parachain) developer understands what is about to happen and why.
2. Second, provide additional detailed code explanation (this you already have).

Reply:
Yeah for this case, I should explain it better. So `add_pallet_runtime` is altogether a different system that I had created during prototypes of pallet engine to add pallets to the CRT macro. What this does is effectively use the structure `RuntimeDeclaration` that's parsed from the source file, append to it with more checks such as indicies, instances, etc. (some of which have no implementation right now but are put off until we dive into handling duplicated entries), and update the tokens for the CRT macro in place in the `output` file. This function is unused here.